src/storage/v2/disk/storage.cpp

```diff
@@ -774,98 +774,283 @@ std::optional<VertexAccessor> DiskStorage::DiskAccessor::FindVertex(storage::Gid
+          guard_vertex.lock();
+        }
+
+        if (!PrepareForWrite(&transaction_, vertex_ptr)) return Error::SERIALIZATION_ERROR;
+        MG_ASSERT(!vertex_ptr->deleted, "Invalid database state!");
+
+        if (opposing_vertex != vertex_ptr) {
+          if (!PrepareForWrite(&transaction_, opposing_vertex)) return Error::SERIALIZATION_ERROR;
+          MG_ASSERT(!opposing_vertex->deleted, "Invalid database state!");
+        }
+
+        edges_collection->pop_back();
 
-    if (ret.GetValue()) {
-      deleted_edges.push_back(*ret.GetValue());
+        auto *from_vertex = reverse_vertex_order ? vertex_ptr : opposing_vertex;
+        auto *to_vertex = reverse_vertex_order ? opposing_vertex : vertex_ptr;
+        mark_edge_as_deleted_func(edge_ref, edge_type, from_vertex, to_vertex);
+
+        CreateAndLinkDelta(&transaction_, vertex_ptr, delta, edge_type, opposing_vertex, edge_ref);
+      }
+
+      return std::make_optional<ReturnType>();
+    };
+
+    // delete the in and out edges from the nodes we want to delete
+    for (auto *vertex_ptr : nodes_to_delete) {
+      auto maybe_error = clear_edges(vertex_ptr, &vertex_ptr->in_edges, Delta::AddInEdgeTag(), false);
+      if (maybe_error.HasError()) {
+        return maybe_error;
+      }
+
+      maybe_error = clear_edges(vertex_ptr, &vertex_ptr->out_edges, Delta::AddOutEdgeTag(), true);
+      if (maybe_error.HasError()) {
+        return maybe_error;
+      }
     }
   }
 
-  std::lock_guard<utils::SpinLock> guard(vertex_ptr->lock);
+  // Detach nodes on the other end, which don't need deletion, by passing once through their vectors
+  {
+    auto detach_non_deletable_nodes = [this, &mark_edge_as_deleted_func](auto *vertex_ptr, auto *edges_collection,
+                                                                         auto &set_for_erasure, auto delta,
+                                                                         auto reverse_vertex_order) {
+      std::lock_guard<utils::SpinLock> guard(vertex_ptr->lock);
+
+      auto mid = std::partition(
+          edges_collection->begin(), edges_collection->end(),
+          [this, &mark_edge_as_deleted_func, &set_for_erasure, vertex_ptr, delta, reverse_vertex_order](auto &edge) {
+            auto [edge_type, opposing_vertex, edge_ref] = edge;
+
+            if (set_for_erasure.contains(edge_ref)) {
+              std::unique_lock<utils::SpinLock> guard;
+              if (config_.properties_on_edges) {
+                auto edge_ptr = edge_ref.ptr;
+                guard = std::unique_lock<utils::SpinLock>(edge_ptr->lock);
+              }
+
+              auto *from_vertex = reverse_vertex_order ? vertex_ptr : opposing_vertex;
+              auto *to_vertex = reverse_vertex_order ? opposing_vertex : vertex_ptr;
+              mark_edge_as_deleted_func(edge_ref, edge_type, from_vertex, to_vertex);
+
+              CreateAndLinkDelta(&transaction_, vertex_ptr, delta, edge_type, opposing_vertex, edge_ref);
```

Comment:
Oh wow, a std::partition with side effects. I mean it works...but not my expected usage.
And I see why you did it, a single pass rather than two.
But predicate functions typically are side-effect free.